runtime/server/repos.go

```diff
@@ -132,14 +136,67 @@ func (s *Server) PutRepoObject(ctx context.Context, req *api.PutRepoObjectReques
 	}
 
 	repoStore, _ := conn.RepoStore()
-	err = repoStore.Put(ctx, repo.ID, req.Path, req.Blob)
+	err = repoStore.PutBlob(ctx, repo.ID, req.Path, req.Blob)
 	if err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
 	return &api.PutRepoObjectResponse{}, nil
 }
 
+func (s *Server) PutRepoObjectFromHTTPRequest(w http.ResponseWriter, req *http.Request, pathParams map[string]string) {
+	ctx := context.Background()
+	if err := req.ParseForm(); err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse request: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	registry, _ := s.metastore.RegistryStore()
+	repo, found := registry.FindRepo(ctx, pathParams["repo_id"])
+	if !found {
+		http.Error(w, "repo not found", http.StatusBadRequest)
+		return
+	}
+
+	conn, err := drivers.Open(repo.Driver, repo.DSN)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to open driver: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	f, _, err := req.FormFile("file")
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse file in request: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	filePath := path.Join("data", pathParams["path"])
```

Comment:
Instead of adding `data` here, I think frontend should uploadt it to "/data/filename". The idea is that repos should basically just be a proxy to a (virtualized or real) filesystem. Then the frontend can act more as an IDE that knows what to do.